File: WOT-iOS/WOT-iOS/core/datasources/WOTTree/WOTNode.m
//
//  WOTNode.m
//  WOT-iOS
//
//  Created by Pavel Yeshchyk on 6/29/15.
//  Copyright (c) 2015 Pavel Yeshchyk. All rights reserved.
//

#import "WOTNode.h"

@interface WOTNode ()
@property (nonatomic, readwrite, strong)NSMutableOrderedSet<WOTNodeProtocol> *childSet;
@end

@implementation WOTNode

@synthesize imageURL;
@synthesize isVisible;
@synthesize name;
@synthesize parent;
@synthesize childList;
@synthesize index;
@synthesize hashString;
@synthesize stepParentRow;
@synthesize stepParentColumn;
@synthesize indexInsideStepParentColumn;
@synthesize relativeRect;

- (NSString *)hashString {

    return [NSString stringWithFormat:@"%d",self.hash];
}

- (void)dealloc {

    [self removeChildren: NULL];
}

- (id)init {
    
    self = [super init];
    if (self){
    
        self.isVisible = YES;
    }
    return self;
}

- (id)copyWithZone:(NSZone *)zone {
    
    WOTNode *result = [[[self class] allocWithZone: zone] init];
    result.isVisible = self.isVisible;
    result.imageURL = self.imageURL;
    result.name = self.name;
    return result;
}

- (id)initWithName:(NSString *)name {
    
    self = [self init];
    if (self){
        
        self.name = name;
    }
    return self;
}

- (id)initWithName:(NSString *)name imageURL:(NSURL *)imageURL{
    
    self = [self initWithName:name];
    if (self){
        
        self.imageURL = imageURL;
    }
    return self;
}

- (NSUInteger)hash {


    return [self.name hash];
}


- (NSArray *)children {
    return [self.childSet array];
}

- (NSString *)fullName {
    
    if (self.parent == nil) {
        
        return self.name;
    } else {
        
        NSString *parentFullName = self.parent.fullName;
        return [NSString stringWithFormat:@"%@.%@",parentFullName,self.name];
    }
    
}

- (void)addChild:(WOTNode * _Nonnull)child {
    
    if (!self.childSet) {
        
        self.childSet = [[NSMutableOrderedSet<WOTNodeProtocol> alloc] init];
    }
    
    child.parent = self;
    [self.childSet addObject:child];
}

- (void)removeChild:(id<WOTNodeProtocol>)child completion:(void (^)(id<WOTNodeProtocol> _Nonnull))completion {

    [self.childSet enumerateObjectsUsingBlock:^(WOTNode *node, NSUInteger idx, BOOL *stop) {

        if (completion) {

            completion(node);
        }
        node.parent = nil;
        [node removeChildren: NULL];

    }];
    [self.childSet removeAllObjects];
}

- (void)removeChildren:(void (^)(id<WOTNodeProtocol> _Nonnull))completion {
    [self.childSet enumerateObjectsUsingBlock:^(WOTNode *node, NSUInteger idx, BOOL *stop) {

        if (completion) {

            completion(node);
        }
        node.parent = nil;
        [node removeChildren: NULL];
    }];
    [self.childSet removeAllObjects];
}

- (void)removeAllNodesWithCompletionBlock:(WOTNodeRemoveCompletionBlock)completionBlock {
    
    [self.childSet enumerateObjectsUsingBlock:^(WOTNode *node, NSUInteger idx, BOOL *stop) {

        if (completionBlock) {
            
            completionBlock(node);
        }
        node.parent = nil;
        [node removeAllNodesWithCompletionBlock:completionBlock];
    }];
    [self.childSet removeAllObjects];
}

- (void) addChildArray:(NSArray<id<WOTNodeProtocol>> *)childArray {
    
    NSSet<WOTNodeProtocol> *set = [NSSet<WOTNodeProtocol> setWithArray:childArray];
    [set enumerateObjectsUsingBlock:^(id  _Nonnull child, BOOL * _Nonnull stop) {
        id<WOTNodeProtocol> node = child;
        node.parent = self;

    }];

    if (!self.childSet) {
        self.childSet =  [[NSMutableOrderedSet<WOTNodeProtocol> alloc] init];
    }

    [self.childSet unionSet: set];
}

- (void)removeChild:(id<WOTNodeProtocol>)child completionBlock:(WOTNodeRemoveCompletionBlock)completionBlock{

    if (completionBlock) {
        
        completionBlock(child);
    }
    child.parent = nil;
    [self.childSet removeObject:child];
}

- (NSOrderedSet *)childList {
    return self.childSet;
}

@end

File: WOT-iOS/WOT-iOS/core/datasources/WOTTree/WOTNode.swift
//
//  WOTNode.swift
//  WOT-iOS
//
//  Created by Pavel Yeshchyk on 6/25/18.
//  Copyright © 2018 Pavel Yeshchyk. All rights reserved.
//

import Foundation
import ObjectiveC

public typealias WOTNodeProtocolRemoveCompletion = (WOTNodeProtocol) -> ()

@objc
public protocol WOTNodeProtocol: NSObjectProtocol, NSCopying {
    var hashString: String { get }
    var name: String { get set }
    var children: [WOTNodeProtocol] { get }
    var endpoints: [WOTNodeProtocol] { get }
    var parent: WOTNodeProtocol? { get set }
    var imageURL: NSURL { get set }
    var childList: Set<AnyHashable> { get }
    var isVisible: Bool { get set }
    var fullName: String { get }
    var index: Int { get set }
    var indexInsideStepParentColumn: Int { get set }
    var stepParentColumn: WOTNodeProtocol? { get set }
    var stepParentRow: WOTNodeProtocol? { get set }

    var relativeRect: NSValue? { get set }

    func addChild(_ child: WOTNodeProtocol)
    func addChildArray(_ childArray: [WOTNodeProtocol])
    func removeChild(_ child: WOTNodeProtocol, completion: WOTNodeProtocolRemoveCompletion)
    func removeChildren(_ completion: WOTNodeProtocolRemoveCompletion?)
}

public class WOTNodeSwift: NSObject, WOTNodeProtocol {

    override public var hashValue: Int {
        return self.fullName.hashValue
    }

    override public var hash: Int {
        return self.fullName.hashValue
    }

    public var hashString: String {
        return String(format: "%d", self.hashValue)
    }

    public var name: String = ""

    public var children: [WOTNodeProtocol] = [WOTNodeProtocol]()

    public var endpoints: [WOTNodeProtocol] {
        return WOTNodeEnumerator.sharedInstance.endpoints(node: self)
    }

    public var parent: WOTNodeProtocol?

    public var imageURL: NSURL = NSURL(string: "mock")!

    public var childList: Set<AnyHashable> = Set<AnyHashable>()

    public var isVisible: Bool = true

    public var fullName: String {
        guard let parent = self.parent else {
            return self.name
        }
        return String(format: "%@.%@", parent.fullName, self.name)
    }

    public var index: Int = 0

    public var indexInsideStepParentColumn: Int = 0

    public var stepParentColumn: WOTNodeProtocol?

    public var stepParentRow: WOTNodeProtocol?

    public var relativeRect: NSValue?

    @objc
    public required init(name nameValue: String) {
        super.init()
        self.name = nameValue
    }

    public func copy(with zone: NSZone? = nil) -> Any {
        let result = type(of: self).init(name: self.name)
        result.isVisible = self.isVisible
        result.imageURL = self.imageURL
        return result;
    }

    public func addChild(_ child: WOTNodeProtocol) {
        child.parent = self
        self.children.append(child)
    }

    public func addChildArray(_ childArray: [WOTNodeProtocol]) {
        childArray.forEach { (child) in
            child.parent = self
            self.children.append(child)
        }
    }

    public func removeChild(_ child: WOTNodeProtocol, completion: (WOTNodeProtocol) -> ()) {
        guard let index = (self.children.index { $0 === child }) else {
            return
        }
        child.parent = nil
        self.children.remove(at: index)
    }

    public func removeChildren(_ completion: WOTNodeProtocolRemoveCompletion?) {
        self.children.forEach { (child) in
            child.removeChildren({ (node) in

                node.parent = nil
            })
            completion?(child)
        }
        self.children.removeAll()
    }
}
